File: src/trees/Diameter.java
package trees;

public class Diameter {

    private class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    int ans = 0;
    public int diameter(TreeNode root) {
        if(root == null){
            return 0;
        }
        int left = diameter(root.left);
        int right = diameter(root.right);
        int maxHeight = 1 + Integer.max(left, right);
        int temp = Integer.max(1 + left + right, maxHeight);
        ans = Integer.max(ans, temp);
        return maxHeight;
    }

    public int diameterOfBinaryTree(TreeNode root){
        if(root == null){
            return 0;
        }
        diameter(root);
        return ans - 1;
    }
}

File: src/Trees/Diameter.kt
package Trees

class Diameter {
    class TreeNode(var `val`: Int) {
        var left: TreeNode? = null
        var right: TreeNode? = null
    }
    var ans = 0
    fun diameterOfBinaryTree(root: TreeNode?): Int {
        fun dfs(rootNew: TreeNode?): Int{
            if (rootNew == null)return 0
            val left = dfs(rootNew.left)
            val right = dfs(rootNew.right)
            val temp = 1 + Math.max(left, right)
            val res = Math.max(temp, 1 + left + right)
            ans = Math.max(ans, res)
            return temp
        }
        return ans - 1
    }
}